File: Shortest Bridge.py
#link: https://leetcode.com/problems/shortest-bridge/description/

from typing import List
from collections import deque


class Solution:

    def valid(self, i: int, j: int, n: int, m: int) -> bool:
        return i >= 0 and j >= 0 and i < n and j < m

    def changeType(self, i: int, j: int, grid: List[List[int]], n: int, m: int) -> None:
        if not self.valid(i, j, n, m) or grid[i][j] == 0 or grid[i][j] == 2:
            return
        grid[i][j] = 2
        for idx in range(4):
            self.changeType(i + self.dx[idx], j + self.dy[idx], grid, n, m)

    def shortestBridge(self, grid: List[List[int]]) -> int:
        self.dx = [-1, 0, 1, 0]
        self.dy = [0, 1, 0, -1]
        n, m = len(grid), len(grid[0])
        ok = True
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    self.changeType(i, j, grid, n, m)
                    ok = 0
                    break
            if not ok:
                break

        q = deque()
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 2:
                    q.append((i, j))
        level = 0
        while len(q) != 0:
            sz = len(q)
            for _ in range(sz):
                cur = q.popleft()
                for idx in range(4):
                    x, y = cur[0] + self.dx[idx], cur[1] + self.dy[idx]
                    if not self.valid(x, y, n, m):
                        continue
                    if grid[x][y] == 0:
                        grid[x][y] = 2
                        q.append((x, y))
                    if grid[x][y] == 1:
                        return level
            level += 1

File: Shortest Bridge.cpp
// link : https://leetcode.com/problems/shortest-bridge/description/

class Solution {
public:
    const int dx[4] = { -1,0,1,0 };
    const int dy[4] = { 0,1,0,-1 };
    inline bool valid(int& i, int& j, int& n, int& m) {
        return (i >= 0 && j >= 0 && i < n&& j < m);
    }
    void changeType(int i, int j, vector<vector<int>>& grid,int&n,int&m) {
        if (!valid(i, j, n, m) || grid[i][j] == 0 || grid[i][j] == 2)
            return;

        grid[i][j] = 2;
        for (int idx = 0; idx < 4; idx++)
            changeType(i + dx[idx], j + dy[idx], grid, n, m);
    }
    int shortestBridge(vector<vector<int>>& grid) {

        int n = grid.size(), m = grid[0].size(), ok = 1;
        for (int r = 0; r < n && ok; r++) {
            for (int c = 0; c < m && ok; c++) {
                if (grid[r][c] == 1) {
                    changeType(r, c, grid, n, m);
                    ok = 0;
                }
            }
        }

        queue<pair<int, int>>q;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 2) {
                    q.push({ i,j });
                }
            }
        }
        int level = 0;
        while (!q.empty()) {
            int sz = q.size();
            while (sz--) {
                pair<int, int>cur = q.front();
                q.pop();
                for (int idx = 0; idx < 4; idx++) {
                    int x = cur.first + dx[idx], y = cur.second + dy[idx];
                    if (!valid(x, y, n, m))
                        continue;

                    if (grid[x][y] == 0) {
                        grid[x][y] = 2;
                        q.push({ x,y });
                    }

                    if (grid[x][y] == 1)
                        return level;
                }
            }
            level++;
        }
        return level;
    }
};